### ISSUE TITLE
Can't continue boot with missing LUKS disk

### ISSUE THREAD

**Issue by ElvishJerricco** [NixOS/nixpkgs#74281]
**Describe the bug**
When booting a NixOS config with an inessential LUKS disk, the boot will fail if the disk is not present. Attempting to continue the boot will cause a kernel panic due to `init` trying to exit.

**To Reproduce**

```
# configuration.nix
{ pkgs, lib, ... }: {
  # Override because virtualisation removes it
  boot.initrd.luks.devices = lib.mkOverride (-1) {
    doesnt-exist.device = "/dev/disk/by-id/doesnt-exist";
  };
  services.mingetty.autologinUser = "root";
  nix.nixPath = ["nixpkgs=${pkgs.path}"];
}
```

```
$ nix run -I nixpkgs=channel:nixos-unstable -f "<nixpkgs/nixos>" --arg configuration ./configuration.nix vm -c run-nixos-vm
```

**Expected behavior**

NixOS should be able to accept the missing disk and attempt to boot without it.

**Actual behavior**

NixOS crashes and fails to boot.

**Screenshots**
![prompt](https://user-images.githubusercontent.com/1365692/69654658-24887f80-1043-11ea-8049-19eaec7523e3.png)

![panic](https://user-images.githubusercontent.com/1365692/69654671-26ead980-1043-11ea-9b5a-77cdd7b1152f.png)

**Additional context**

For example, in a ZFS pool with all disks encrypted with LUKS, it's reasonable to assume ZFS will function with one disk missing. I'm thinking this should probably be an option, similar to `fileSystems."/foo".options = ["nofail"];`.

**Metadata**
Ran `nix-info -m` in the VM after commenting out LUKS device.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.84, NixOS, 20.03pre202088.e89b21504f3 (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.1`
 - nixpkgs: `/nix/store/99n1237y0lm1vn8lks0rggnp1armhdr4-a46m63g89mx10waxxp71c4yp6jwcfkd4-nixexprs.tar.xz`
```

Maintainer information:
```yaml
# a list of nixos modules affected by the problem
module:nixos/modules/system/boot/luksroot.nix
```

**Comment by github-kjsj** [comment]
To add a second (albeit neophyte when it comes to NixOS) voice: I've run into this while using LVMRAID on LUKS, which this makes nonviable if we want to be able to survive a missing or dead device on boot.

**Comment by PAI5REECHO** [comment]
bad bot

**Comment by 1nv0k32** [comment]
+1 here. One the system users has their partition on a portable disk encrypted with Luks. Not having the drive, causes system to be unusable.

**Comment by ElvishJerricco** [comment]
FWIW, with `boot.initrd.systemd.enable = true;`, you actually can make the failure acceptable with `boot.initrd.luks.devices.<name>.crypttabExtraOpts = ["nofail"];` because systemd initrd uses `systemd-cryptsetup-generator`.

Similarly, it's actually kind of a mistake that NixOS's LUKS options always cause devices to be unlocked in initrd. There's no easy way to do this right, since there's no easy way to tell which devices are required for which file systems. But the point is that the sysadmin *should* be putting their LUKS devices into `environment.etc."crypttab"` when it's possible to decrypt them in stage 2 instead of initrd. That way `systemd-cryptsetup-generator` can cause the devices to be unlocked during boot with proper systemd unit ordering/dependencies.

**Comment by someplaceguy** [comment]
@ElvishJerricco Thanks for the insight. I wasn't aware of the different booting procedure with `boot.initrd.systemd.enable = true`, so I have a few questions about using systemd in initrd that I was hoping you could answer:

1. Assuming I have a ZFS pool that can be unlocked in stage 2 and I add the corresponding LUKS devices to `/etc/crypttab` like you say, does `systemd-cryptsetup-generator` automatically add the proper LUKS unlocking dependencies to the `zfs-import-<poolname>.service` unit that runs when booting? (I'm guessing it doesn't, since systemd is probably not aware of ZFS?)

2. If I have a ZFS pool that contains the root filesystem, judging by your comment I'm assuming that adding its devices to `/etc/crypttab` wouldn't be useful because the root filesystem needs to be mounted in stage 1, right?

3. Also, in the same scenario as the point above, I'm assuming that even when using systemd in initrd, if one of the redundant LUKS devices in the ZFS pool with the root filesystem fails to unlock (e.g. because the disk is missing or malfunctioning) then the boot process would still fail even though the ZFS pool would still work fine (although in "degraded" mode) if the boot process were to continue with the missing LUKS device?

**Comment by ElvishJerricco** [comment]
@someplaceguy 

> Assuming I have a ZFS pool that can be unlocked in stage 2 and I add the corresponding LUKS devices to /etc/crypttab like you say, does systemd-cryptsetup-generator automatically add the proper LUKS unlocking dependencies to the zfs-import-<poolname>.service unit that runs when booting? (I'm guessing it doesn't, since systemd is probably not aware of ZFS?)

The generator doesn't know anything about ZFS so no, it won't do that. However, [this PR](https://github.com/NixOS/nixpkgs/pull/227208) fixes that by adding `After=cryptsetup.target` to ZFS import services. Unfortunately that's not merged yet.

> If I have a ZFS pool that contains the root filesystem, judging by your comment I'm assuming that adding its devices to /etc/crypttab wouldn't be useful because the root filesystem needs to be mounted in stage 1, right?

Right. Can't mount stage 2 before stage 2 :P

> Also, in the same scenario as the point above, I'm assuming that even when using systemd in initrd, if one of the redundant LUKS devices in the ZFS pool with the root filesystem fails to unlock (e.g. because the disk is missing or malfunctioning) then the boot process would still fail even though the ZFS pool would still work fine (although in "degraded" mode) if the boot process were to continue with the missing LUKS device?

NixOS actually does try to import a pool degraded after 60 seconds. So if some disks don't appear it still imports degraded.

**Comment by someplaceguy** [comment]
Ok, so if I understand you correctly, then this issue can be worked around completely if you do all of the following steps:

1. You set `boot.initrd.systemd.enable = true`.
2. You set `boot.initrd.luks.devices.<name>.crypttabExtraOpts = ["nofail"]` for those LUKS devices that are not strictly required to be unlocked for the boot to proceed, either because they don't contain the root filesystem, or because even though they contain the root filesystem, the filesystem can still be mounted if the LUKS device is missing (e.g. LUKS devices used in RAID-1 / RAID-5 systems or analogous btrfs / ZFS pools).
3. Either you're not a ZFS user or PR #227208 is merged, so that ZFS pools can be imported correctly when using systemd in initrd.

Furthermore, you can optionally create `environment.etc."crypttab"` manually instead of using `boot.initrd.luks.devices` for those LUKS devices that are not needed for the root filesystem.

I assume the latter would allow the boot process to (potentially) finish more quickly, because LUKS device opening/unlocking can start/run in parallel during boot, due to the unlocking being split into separate units in systemd.

Thanks @ElvishJerricco, this is awesome.

**Comment by ElvishJerricco** [comment]
For posterity's sake, #227208 has been edited to not include any LUKS related changes

**Comment by koppor** [comment]
For the googlers who land here, because the cannot boot at all. (For me, it was changing the file system of an ext4 partition to f2fs. Then I could not boot)

1. Log in into rescue mode using the root password
2. `cd /etc`
3. `mv fstab fstab.old`
4. Follow the symlinks of `fstab` and copy that file as `fstab`
   - `ls -la /etc/static/fstab`
   - Check that link. Let is be `X`
   - `cp X fstab`
5. Edit `fstab` according to your needs. E.g., remove the non-mountable drives
6. Make fstab immutable `chattr +i fstab`
7. Reboot. If it does not work, go back to step 5.
8. Make fstab mutable again: `chattr -i fstab`
9. Boot into nix and try to fix the issue. Maybe using the hint above - https://github.com/NixOS/nixpkgs/issues/74281#issuecomment-1646272631

**Comment by ElvishJerricco** [comment]
@koppor for the record, I don't think that's related to the issue here at all.

**Comment by grulles** [comment]
Hello, I'm coming ! 
I read this post and I have pretty the same trouble. 
I had successfully implement LUKS / VG / LV mirror / and two efiboot every thing manage by DISKO
I also add  crypttabExtraOpts = ["nofail"];
If my sys boot with 2 disks : OK no issue
If I remove one disk, the sys can't boot because LUSK device is missing and VG01 can't be mounted. 
Do you know what I missed in my conf ? 
I can provide more details if needed.
Thanks for your help !

**Comment by Tillerino** [comment]
This is too bad. I tried to add `crypttabExtraOpts = ["nofail"]` with `boot.initrd.systemd.enable`, but unfortunately I am also using `boot.initrd.luks.reusePassphrases` and apparently these are mutually exclusive:

https://github.com/NixOS/nixpkgs/blob/ca534a76c4afb2bdc07b681dbc11b453bab21af8/nixos/modules/system/boot/luksroot.nix#L1095-L1098
